pipeline_tools/http_requests.py

```diff
@@ -0,0 +1,272 @@
+        self.retry_multiplier = self._get_param(RETRY_MULTIPLIER, '1', float)
+        self.retry_max_interval = self._get_param(RETRY_MAX_INTERVAL, '10', float)
+
+    def _get_param(self, name, default, convert_fn=None):
+        param_str = os.environ.get(name, default)
+        # handle case where environment var is set to empty string
+        if not param_str:
+            param_str = default
+
+        if convert_fn is not None:
+            return convert_fn(param_str)
+        else:
+            return param_str
+
+    def get(self, *args, **kwargs):
+        """Calls requests.get function.
+        
+        In addition to calling requests.get, this function will record the request
+        and response if the HttpRequests object's should_record attribute is True.
+
+        Args:
+            All arguments that requests.get accepts are accepted here.
+
+        Returns:
+            A requests.Response object representing the response to the request.
+
+        Raises:
+            ValueError if response status code indicates an error.
+        """
+        kwargs['method'] = 'get'
+        return self._http_request_with_retry(*args, **kwargs)
+
+    def put(self, *args, **kwargs):
+        """Calls requests.put function.
+        
+        In addition to calling requests.put, this function will record the request
+        and response if the HttpRequests object's should_record attribute is True.
+
+        Args:
+            All arguments that requests.put accepts are accepted here except
+            for 'body' -- use 'json' instead.
+
+        Returns:
+            A requests.Response object representing the response to the request.
+
+        Raises:
+            ValueError if response status code indicates an error.
+        """
+        kwargs['method'] = 'put'
+        return self._http_request_with_retry(*args, **kwargs)
+
+    def post(self, *args, **kwargs):
+        """Calls requests.post function.
+        
+        In addition to calling requests.post, this function will record the request
+        and response if the HttpRequests object's should_record attribute is True.
+
+        Args:
+            All arguments that requests.post accepts are accepted here except
+            for 'body' -- use 'json' instead.
+
+        Returns:
+            A requests.Response object representing the response to the request.
+
+        Raises:
+            ValueError if response status code indicates an error.
+        """
+        kwargs['method'] = 'post'
+        return self._http_request_with_retry(*args, **kwargs)
+
+    def _http_request_with_retry(self, *args, **kwargs):
+        def _is_retryable_status_code(error):
+            if not isinstance(error, requests.HTTPError):
+                return False
+            return not (400 <= error.response.status_code <= 499)
+
+        if 'retry' in kwargs:
+            retry = kwargs['retry'] | retry_if_exception(_is_retryable_status_code)
```

Comment:
If I understand it correctly, does the `kwargs['retry']` here used for keeping the ability to tweak/fine-tune retry strategy for each individual functions for us? If this is the case, strong 👍 for it. My concern about putting retry parameters into env variables is that we lose the ability to tweak the retry strategy (e.g. retry timeout seconds) for various individual functions.

Reply:
How about including the parameters in the class constructor to allow optionally overriding the environment variables?

Reply:
`kwargs['retry']` lets you override the way we decide whether to retry. By default, we retry only if there is a 5xx error, but we use the `kwargs['retry']` argument in a couple of places to let us retry based on the content of the returned json.

That's a good suggestion @samanehsan, but let's wait to add that ability until we really want to use it somewhere.

I'm hoping that we can switch from using environment variables to command line parameters in a future refactoring that was too big to undertake in this already big PR, so hopefully this situation is temporary anyway.